include/rspq.h

```diff
@@ -0,0 +1,704 @@
+ *       that freeing a block is only done when no other blocks reference it.
+ */
+void rspq_block_free(rspq_block_t *block);
+
+/**
+ * @brief Start building a high-priority queue.
+ * 
+ * This function enters a special mode in which a high-priority queue is
+ * activated and can be filled with commands. After this function has been
+ * called, all commands will be put in the high-priority queue, until
+ * #rspq_highpri_end is called.
+ * 
+ * The RSP will start processing the high-priority queue almost instantly
+ * (as soon as the current command is done), pausing the normal queue. This will
+ * also happen while the high-priority queue is being built, to achieve the
+ * lowest possible latency. When the RSP finishes processing the high priority
+ * queue (after #rspq_highpri_end closes it), it resumes processing the normal
+ * queue from the exact point that was left.
+ * 
+ * The goal of the high-priority queue is to either schedule latency-sensitive
+ * commands like audio processing, or to schedule immediate RSP calculations
+ * that should be performed right away, just like they were preempting what
+ * the RSP is currently doing.
+ * 
+ * It is possible to create multiple high-priority queues by calling
+ * #rspq_highpri_begin / #rspq_highpri_end multiples time with short
+ * delays in-between. The RSP will process them in order. Notice that
+ * there is a overhead in doing so, so it might be advisable to keep
+ * the high-priority mode active for a longer period if possible. On the
+ * other hand, a shorter high-priority queue allows for the RSP to
+ * switch back to processing the normal queue before the next one
+ * is created.
+ * 
+ * @note It is not possible to create a block while the high-priority queue is
+ *       active. Arrange for constructing blocks beforehand.
+ *       
+ * @note It is currently not possible to call a block from the
+ *       high-priority queue. (FIXME: to be implemented)
+ *       
+ */
+void rspq_highpri_begin(void);
+
+/**
+ * @brief Finish building the high-priority queue and close it.
+ * 
+ * This function terminates and closes the high-priority queue. After this
+ * command is called, all following commands will be added to the normal queue.
+ * 
+ * Notice that the RSP does not wait for this function to be called: it will
+ * start running the high-priority queue as soon as possible, even while it is
+ * being built.
+ */
+void rspq_highpri_end(void);
+
+/**
+ * @brief Wait for the RSP to finish processing all high-priority queues.
+ * 
+ * This function will spin-lock waiting for the RSP to finish processing
+ * all high-priority queues. It is meant for debugging purposes or for situations
+ * in which the high-priority queue is known to be very short and fast to run,
+ * so that the overhead of a syncpoint would be too high.
+ * 
+ * For longer/slower high-priority queues, it is advisable to use a #rspq_syncpoint_t
+ * to synchronize (though it has a higher overhead).
```

Comment:
So why it is advisable to use it instead? If we only want to wait for high prio to end is it better to use a syncpoint as last command or `rspq_highpri_sync`?

Reply:
rspq_highpri_sync will spin loop. If you know it's going to take a long time, it might be better to add a syncpoint as last command in the highpri queue, go doing something else, and later check whether it has finished or not. Obviously you could instead call rspq_highpri_sync later -- if you know that no other highpri stuff has happened.

Reply:
Actually is it even possible to use syncpoints in highpri? Because they might then be triggered out of sequence?

Reply:
oh sorry that's right, you can't do that and is probably also covered by an assert. So this comment should be removed.

Reply:
It wasn't covered by an assert so far, so I added one.

Reply:
I actually forgot to remove the comment, so I've done that now :)